### PR TITLE
Handling Float API errors gracefully

### PR DESCRIPTION
**Pull request by ccostino**
Fixes issue #669 

This changeset provides much more robust error handling for the Float API calls and more graceful handling of them so that users can still see and enter their timecards.  If an error occurs while accessing Float data, the user is presented with a message saying so and the error is logged behind the scenes so we can see what went wrong and why.  We currently only see a stack trace saying that we cannot call a method on a NoneType object.

There's more to be done here for the future, but at this point we just need to make sure the timecards always work regardless of whether or not retrieving the Float data does and, in the case of such errors, where the breakdown actually happened and why.

### CODE REVIEW

File: tock/hours/float.py [hunk 1]
```diff
@@ -13,43 +17,76 @@
...
-        timeoffs=float_timeoffs,
-        work_days=work_days
-    )
+    try:
+        float_tasks = get_float_tasks(
+            start_date=rp.start_date,
+            float_people_id=float_people_id,
+            weeks=1
+        )
+    except:
```
Comment (tbaxter-18f) [inline tock/hours/float.py]: Did you think about handling the exception a bit more and passing it through? something along the lines of 

```
except Exception as e:
    logger.error(e)
```

I think that's right. If not, it's close.

File: tock/hours/float.py [hunk 2]
```diff
@@ -13,43 +17,76 @@
...
+                start_date=rp.start_date,
+                end_date=rp.end_date,
+                tasks=float_tasks,
+                work_days=work_days,
+                work_hours=work_hours
+            )
+        except:
+            logger.error(sys.exc_info()[0])
+
+    return tasks
```
Comment (tbaxter-18f) [inline tock/hours/float.py]: Would you want to return some feedback that there were errors?

File: tock/tock/templates/hours/timecard_form.html [hunk 3]
```diff
@@ -11,7 +11,7 @@
     </p>
   </div>
 </div>
-{% if float_data %}
+{% if float_data is not None %}
```
Comment (tbaxter-18f) [inline tock/tock/templates/hours/timecard_form.html]: Why was this change needed?

File: tock/tock/templates/hours/timecard_form.html [hunk 4]
```diff
@@ -28,7 +28,13 @@
            {% if i.task_name %}
             assigned to {{ i.task_name }}
            {% endif %}
-        </li>
+         </li>
+        {% empty %}
+         <li>We're sorry, there was an error retrieving your data from Float.
```
Comment (tbaxter-18f) [inline tock/tock/templates/hours/timecard_form.html]: We could, if we wanted, surface the error here. I can see where that might not be desirable, though. Just a thought.

File: tock/hours/float.py (hunk 1 above)
Comment (tadhg-ohiggins) [reply tock/hours/float.py]: I agree, we should be catching specific exception types only.

Comment (ccostino) [reply tock/hours/float.py]: You're right in that this isn't ideal, but at the moment we don't necessarily know what the error will be, or at least I'm not 100% confident we will (by and large I'm seeing `attempting to call a method on NoneType objects`, but again, no guarantee).  As far as the handling itself goes, that may be better, let me give it a shot!

File: tock/hours/float.py (hunk 2 above)
Comment (ccostino) [reply tock/hours/float.py]: See answer to the template change. :-)

File: tock/tock/templates/hours/timecard_form.html (hunk 3 above)
Comment (ccostino) [reply tock/tock/templates/hours/timecard_form.html]: This was needed because there is a state where `float_data` is `None` and that's valid - in that case we don't want to display the Float information box at all.  What this change does is allows me to send an empty list instead and check for that - if there's no data, I can assume there was an error retrieving the information and provide a message.

File: tock/tock/templates/hours/timecard_form.html (hunk 4 above)
Comment (ccostino) [reply tock/tock/templates/hours/timecard_form.html]: Displaying the error here would be useless as this is user facing; folks aren't going to know what it means.  It's on us as the devs/admins to look in the logs and debug from there. :-)

Comment (ccostino) [reply tock/tock/templates/hours/timecard_form.html]: Correction:  displaying the raw error here.

File: tock/tock/templates/hours/timecard_form.html (hunk 3 above)
Comment (tbaxter-18f) [reply tock/tock/templates/hours/timecard_form.html]: I think -- but I could be wrong -- that since Django templates are more truthy, an empty list still evaluates to none. So the prior way should still work. But it's probably not worth changing it back, either.

Comment (ccostino) [reply tock/tock/templates/hours/timecard_form.html]: Ah I'll have to check!  That said, this is also more explicit. :-)

**Comment by tbaxter-18f** [comment]
Looks like it meets the immediate need to me.

File: tock/hours/float.py (hunk 1 above)
Comment (ccostino) [reply tock/hours/float.py]: And to add a bit more color here:  the handling of the Float API calls in general could be improved quite a bit.  Again, for the sake of time and stability, this is just to shore things up for the interim.

**Comment by ccostino** [comment]
Thank you for taking a look so quickly, @tbaxter-18f and @tadhg-ohiggins!
